### ISSUE TITLE
DataFrame (tabular data viewer)

### ISSUE THREAD

**Issue by xiaodaigh** [julia-vscode/julia-vscode#437]
Would be awesome to have some native tabular data viewer builtin e.g.

https://marketplace.visualstudio.com/items?itemName=GrapeCity.gc-excelviewer

The one feature I miss from Rstudio is being able to view a data.frame in the editor/IDE

**Comment by davidanthoff** [comment]
Yes, very much on my radar. I've played a little bit with some code, and am pretty optimistic that I can get a first version out fairly easily. No promises about a timeline, but it will happen.

My current thinking is that one would have to explicitly ask to display a table in the VS Code UI, and not make this the default at the REPL. So
````julia
julia> df
````
by itself would still show the content in the REPL, but something like
````julia
julia> grid(df)
````
would show it in the UI. I would also integrate that with the piping from the Queryverse world:
````julia
julia> df |> grid()
````
And it would of course work with any [IterableTables.jl](https://github.com/davidanthoff/IterableTables.jl), not just [DataFrames.jl](https://github.com/JuliaData/DataFrames.jl).

Does that broadly make sense to you?

**Comment by davidanthoff** [comment]
https://github.com/JuliaEditorSupport/julia-vscode/pull/438 tracks what I'm doing there. Not much to see so far, though.

**Comment by xiaodaigh** [comment]
Ok. Wonder if it's possible to make VSCode Julia more like a full IDE so can view a list of objects and clicking on them also bring up the table.

I had some experience using the https://github.com/bvaughn/react-virtualized to show only part of the table; so it would be awesome addition as performance can serious degradation if not careful with the implementation especially when dealing with large datasets.

**Comment by davidanthoff** [comment]
We had a workspace view in a beta build that showed the list of active global vars, but it still needs some work before we can release it.

I'll take a look at the react grid you tried, thanks for the pointer!

The more immediate issue with large datasets will be how we get the data from the julia process into the node/vscode process without too much overhead... I don't have a good idea right now, my current plan would be to just serialize the data as JSON and then ship over, which is clearly going to be terrible for large datasets... I have the same issue over at https://github.com/davidanthoff/DataVoyager.jl, and I think https://github.com/sglyon/PlotlyJS.jl also has an open issue about this, so a general solution would be really great. I think long term the right approach would be to jump on the apache arrow train for this kind of data transfer between processes, but that would require a lot of low-level infrastructure first.

**Comment by xgdgsc** [comment]
From a glimpse of Atom.jl and atom-julia-client I think it is possible to request data by index ranges and get only a subset rendered to json each time from the julia process with buttons of previous/next and option to jump to page numbers (like the dataframe view of Rstudio).

**Comment by davidanthoff** [comment]
Agreed, it would be nice to have dynamic loading of data. We have to be careful to use the VS Code communication tools between the webview and the Julia process, otherwise it won't work in the remote scenario, though. That shouldn't be a problem, just something to keep in mind.

BUT, I'm actually going to close this issue because the basic functionality to display tables is implemented.
